### ISSUE TITLE
Replace AdviceProvider with Host interface

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#1073]
As mentioned in https://github.com/0xPolygonMiden/miden-vm/pull/1069#issuecomment-1717088131, it could be a good idea to replace the current [AdviceProvider](https://github.com/0xPolygonMiden/miden-vm/blob/next/processor/src/advice/mod.rs#L42) with a more general `Host` interface. This host interface would assume the responsibilities of the current `AdviceProvider` but would also have additional responsibilities like capturing events, debug info, and probably other things.

One simple way to do this is to rename the current `AdviceProvider` into `Host` and add a few more methods to it. This will work, but might make the interface a bit unwieldy.

Another option is to use it as a wrapper around more specialized traits. For example:
```Rust
pub trait Host {
    type AdviceProvider: AdviceProvider;
    type DebugHandler: DebugHandler;
    type SignalHandler: SignalHandler;

    fn adv(&self) -> &Self::AdviceProvider;
    fn adv_mut(&mut self) -> &mut Self::AdviceProvider;

    fn debug(&mut self) -> &Self::DebugHandler;
    fn signal(&mut self) -> &Self::SignalHandler;
}
```
This should also work, but also may be "too generalized" and could be a bit awkward to work with. So, I'm wondering if there is a better way to define the `Host` trait.

**Comment by frisitano** [comment]
I think this is a nice idea.  I'm thinking that maybe we can provide better abstraction by redefining the interface.  We treat the `Host` as it's own virtual machine of sorts against which we can execute `HostFunction`s (requests), we can then retrieve the results from the `Host` by popping items of it's stack.  The `Process` does not have any knowledge about how the host is implemented.  The `Host` trait would look something like this. 

```rust
trait Host {
    type HostFunction;
    type HostError;

    fn execute_host_function(
        &mut self,
        process: &Process<Self>,
        function: Self::HostFunction,
    ) -> Result<(), Self::HostError>;

    fn pop_stack(&mut self) -> Result<Felt, Self::HostError>;

    fn pop_stack_word(&mut self) -> Result<Word, Self::HostError>;

    fn pop_stack_dword(&mut self) -> Result<[Word; 2], Self::HostError>;
}
```

It would then be the responsibility of the object that implements the `Host` trait to handle the host functions, specifically the handlers implemented [here](https://github.com/0xPolygonMiden/miden-vm/tree/main/processor/src/decorators).  This keeps the logic for host functions encapsulated in the struct which is modified by them (the `Host`).  Additionally we encapsulate the logic around signals and any other host specific feature encapsulated in the `Host`.   

I'm thinking it could look something like this:

```rust
enum HostFunction {
    Advice(AdviceInjector),
    Debug(DebugOptions),
}

enum HostError {
    Advice(AdviceError),
    Debug(DebugError),
    Signal(SignalError),
}

struct DefaultHost {
    adv_provider: MemAdviceProvider,
    debug_handler: DebugHandler,
    signal_handler: SignalHandler,
}

impl Host for DefaultHost {
    type HostFunction = HostFunction;
    type HostError = HostError;

    fn execute_host_function(
        &mut self,
        process: &Process<Self>,
        function: Self::HostFunction,
    ) -> Result<(), Self::HostError> {
        match function {
            HostFunction::Advice(injector) => self
                .adv_provider
                .handle_advice_injector(self, process, injector)
                .map_err(HostError::Advice),
            HostFunction::Debug(options) => self
                .debug_handler
                .handle_debug(self, process, options)
                .map_err(HostError::Debug),
        }
    }
}
```

**Comment by bobbinth** [comment]
I like this approach! I do have a few thoughts/comments:

### Comment 1
One potential complication I see is significant additional complexity of implementing new advice providers. Right now, advice providers are pretty "dump" - i.e., most of the actual logic lives in decorator handlers. But if this logic is moved into the advice provider, I wonder how much more effort it would be for someone to implement a new advice provider. Maybe there is a good way to encapsulate this logic somehow so that if I wanted to implement a new advice provider, I wouldn't need to re-implement (or copy-paste) this logic.

### Comment 2
Some functions of the advice provider interface currently return various values (e.g., `get_merkle_path`, `update_merkle_node`, `find_lone_leaf`, and a few others). I guess we can put these values onto the advice stack and then pop them off, but this creates a few inconveniences:

1. This is going to be slightly less efficient - though, with the amounts of data we are handling here, this is probably negligible.
2. The interface becomes somewhat untyped (i.e., if we want to change the return type some of these functions, the compiler will not help us). But maybe given the small number of such functions and with proper tests, this is OK.
3. In some cases, we may not know how many items to pop off the advice stack. For example, if we are getting back a list of elements for some key form the advice map, we don't really know how many elements will be returned.

To address the 3rd point above, we could return the number of items pushed onto the advice stack from `execute_host_function` function. This way, the signature would become something like:
```Rust
fn execute_host_function(
    &mut self,
    process: &Process<Self>,
    function: Self::HostFunction,
) -> Result<usize, Self::HostError>;
```

For a related reason, we might want to add an extra method to pop a vector off of the advice stack. Something like:
```Rust
fn pop_stack_vec(&mut self, len: usize) -> Result<Vec<Felt>, Self::HostError>;
```

### Comment 3
I'm not sure `HostFunction` can be an associated type. The VM has specific instructions which invoke specific advice injectors, and thus, the list of `HostFunctions` must be known to the VM ahead of time.

For a similar reason, I'm am not sure we should make error type into an associated type either (might make `ExecutionError` more complicated).

### Comment 4
I would not like to make the `Process` public. One way to get around this is to define a trait which would expose certain functions needed to examine the process state. For example, something like:
```Rust
pub trait ProcessState {
    fn stack(&self) -> &[Felt, 16]
    ...
}
```
Then we can implement `ProcessState` on `Process` and provide this to the host interface. Overall, putting all of the above comments together, the host interface may look like this:

```Rust
trait Host {
    fn execute_host_function<S: ProcessState>(
        &mut self,
        process: &S,
        function: HostFunction,
    ) -> Result<usize, HostError>;

    fn pop_stack(&mut self) -> Result<Felt, HostError>;

    fn pop_stack_word(&mut self) -> Result<Word, HostError>;

    fn pop_stack_dword(&mut self) -> Result<[Word; 2], HostError>;

    fn pop_stack_vec(&mut self, len: usize) -> Result<Vec<Felt>, HostError>;
}

```

I think the key question is still: can the advice injector logic be encapsulated so as not to complicate new implementations of advice providers too much.

**Comment by frisitano** [comment]
> I like this approach! I do have a few thoughts/comments:
> 
> ### Comment 1
> One potential complication I see is significant additional complexity of implementing new advice providers. Right now, advice providers are pretty "dump" - i.e., most of the actual logic lives in decorator handlers. But if this logic is moved into the advice provider, I wonder how much more effort it would be for someone to implement a new advice provider. Maybe there is a good way to encapsulate this logic somehow so that if I wanted to implement a new advice provider, I wouldn't need to re-implement (or copy-paste) this logic.

Good catch, I think the ambition of making it easy to implement new `Host`s and `AdviceProvider`s is foundational.  One way we can do this is by implementing all of the default host function logic in the `Host` trait.  I have provided a proposal at the end of this comment that addresses the various points you raised. 

Just a point on terminology here, for the sake of understanding I would say "the `Host` has an `AdviceProvider`" and not "the `Host` is an `AdviceProvider`".  Sorry about being pedantic but I think it will make communication more efficient.

> ### Comment 2
> Some functions of the advice provider interface currently return various values (e.g., `get_merkle_path`, `update_merkle_node`, `find_lone_leaf`, and a few others). I guess we can put these values onto the advice stack and then pop them off, but this creates a few inconveniences:
> 
> 1. This is going to be slightly less efficient - though, with the amounts of data we are handling here, this is probably negligible.
> 2. The interface becomes somewhat untyped (i.e., if we want to change the return type some of these functions, the compiler will not help us). But maybe given the small number of such functions and with proper tests, this is OK.
> 3. In some cases, we may not know how many items to pop off the advice stack. For example, if we are getting back a list of elements for some key form the advice map, we don't really know how many elements will be returned.
> 
> To address the 3rd point above, we could return the number of items pushed onto the advice stack from `execute_host_function` function. This way, the signature would become something like:
> 
> ```rust
> fn execute_host_function(
>     &mut self,
>     process: &Process<Self>,
>     function: Self::HostFunction,
> ) -> Result<usize, Self::HostError>;
> ```
> 
> For a related reason, we might want to add an extra method to pop a vector off of the advice stack. Something like:
> 
> ```rust
> fn pop_stack_vec(&mut self, len: usize) -> Result<Vec<Felt>, Self::HostError>;
> ```

I would agree this is quite undesirable.  We should try and leverage the compiler guarantees and avoid any unnecessary allocations.  One way in which we can do this is by introducing a `HostResult` enum.  This would look like:

```rust
pub enum HostResult {
    MerklePath(MerklePath),
    DoubleWord([Word; 2]),
    Word(Word),
    Element(Felt),
    Unit,
}
```

> ### Comment 3
> I'm not sure `HostFunction` can be an associated type. The VM has specific instructions which invoke specific advice injectors, and thus, the list of `HostFunctions` must be known to the VM ahead of time.
> 
> For a similar reason, I'm am not sure we should make error type into an associated type either (might make `ExecutionError` more complicated).

Makes sense - we will have these as concrete types.

> ### Comment 4
> I would not like to make the `Process` public. One way to get around this is to define a trait which would expose certain functions needed to examine the process state. For example, something like:
> 
> ```rust
> pub trait ProcessState {
>     fn stack(&self) -> &[Felt, 16]
>     ...
> }
> ```
> 
> Then we can implement `ProcessState` on `Process` and provide this to the host interface.

I like this proposal.  With everything discussed above I think it could look something like this:

```rust
// HOST TYPES
// ================================================================================================
pub enum HostError {
    Advice,
    Debug,
    Signal,
}

pub enum HostResult {
    MerklePath(MerklePath),
    DoubleWord([Word; 2]),
    Word(Word),
    Element(Felt),
    Unit,
}

pub enum HostFunction {
    Advice(AdviceInjector),
    Debug(DebugOptions),
}

// // HOST TRAIT
// // ================================================================================================

pub trait DebugHandler{
   ...
}
pub trait ProcessState {
    fn stack(&self) -> &Stack;
    ...
}

trait Host {
    type AdviceProvider: AdviceProvider;
    type DebugHandler: DebugHandler;

    fn execute_host_function<S: ProcessState>(
        &mut self,
        process: &S,
        function: HostFunction,
    ) -> Result<HostResult, HostError> {
        match function {
            HostFunction::CopyMerkleNode => self.copy_merkle_node_to_adv_stack(process),
           ...
        }
    }

    fn adv_provider(&self) -> Self::AdviceProvider;
    fn mut_adv_provider(&self) -> &mut Self::AdviceProvider;
    fn debug_handler(&self) -> Self::DebugHandler;
    fn mut_debug_handler(&mut self) -> &mut Self::DebugHandler;

    fn copy_merkle_node_to_adv_stack<S: ProcessState>(&mut self, process: &S) -> Result<HostResult, HostError> {
        // read node depth, node index, and tree root from the stack
        let depth = process.stack().get(0);
        let index = process.stack().get(1);
        let root = [process.stack().get(5), process.stack().get(4), process.stack().get(3), process.stack().get(2)];

        // look up the node in the advice provider
        let node = self.adv_provider().get_tree_node(root, &depth, &index)?;

        // push the node onto the advice stack with the first element pushed last so that it can
        // be popped first (i.e. stack behavior for word)
        self.mut_adv_provider().push_stack(AdviceSource::Value(node[3]))?;
        self.mut_adv_provider().push_stack(AdviceSource::Value(node[2]))?;
        self.mut_adv_provider().push_stack(AdviceSource::Value(node[1]))?;
        self.mut_adv_provider().push_stack(AdviceSource::Value(node[0]))?;

        Ok(HostResult::Unit)
    }

    ...
}

// DEFAULT HOST IMPLEMENTATION
// ================================================================================================

struct DefaultHost<A, D> {
    adv_provider: A,
    debug_handler: D,
}

impl<A: AdviceProvider, D: DebugHandler> Host for DefaultHost<A, D> {
    type AdviceProvider =  A;
    type DebugHandler = D;

    fn adv_provider(&self) -> Self::AdviceProvider {
        self.adv_provider
    }

    fn mut_adv_provider(&self) -> &mut Self::AdviceProvider {
        &mut self.adv_provider
    }

    fn debug_handler(&self) -> Self::DebugHandler {
        self.debug_handler
    }

    fn mut_debug_handler(&mut self) -> &mut Self::DebugHandler {
        &mut self.debug_handler
    }

    // Optional: the implementer can override default behaviour for host functions
    fn copy_merkle_node_to_adv_stack<S: ProcessState>(&mut self, process: &S) -> Result<HostResult, HostError> {
        todo!()
    }
}
```

I don't think the `SignalHandler` should be part of the `DefaultHost`.  We should implement a `Host` in `miden-base` which has the `SignalHandler` as it doesn't seem required for the default implementation.

**Comment by frisitano** [comment]
> First, I think the advice provider part can be tackled by structuring the `AdviceProvider` interface as follows:
> 
> ```rust
> pub trait AdviceProvider {
> 
>     // PROVIDED MEthODS
>     // --------------------------------------------------------------------------------------------
> 
>     fn apply_injector<S: ProcessState>(
>         &mut self,
>         process: &S
>         injector: AdviceInjector,
>     ) -> Result<usize, HostError> {
>         match injector {
>             AdviceInjector::HdwordToMap { domain } => self.insert_hdword_into_adv_map(process, *domain),
>             ...
>         }
>     }
> 
>     fn insert_hdword_into_adv_map<S: ProcessState>(
>         &mut self,
>         process: &S
>         domain: Felt,
>     ) -> Result<(), ExecutionError> {
>         // TODO: build key and values from the data in the process
> 
>         // invoke a required method
>         self.insert_into_map(key, values)
>     }
> 
>     // REQUIRED METHODS
>     // --------------------------------------------------------------------------------------------
> 
>     fn insert_into_map(&mut self, key: Word, values: Vec<Felt>) -> Result<(), HostError>
> }
> ```

I like this proposal and think it will work well.  One point on using traits with default implementations is that if the trait is public then all methods associated with the trait will be public as well and as such part of the public interface.

> Then, we can keep the `Host` interface largely the same as before:
> 
> ```rust
> trait Host {
>     fn execute_host_function<S: ProcessState>(
>         &mut self,
>         process: &S,
>         function: HostFunction,
>     ) -> Result<usize, HostError>;
> 
>     fn pop_stack(&mut self) -> Result<Felt, HostError>;
> 
>     fn pop_stack_word(&mut self) -> Result<Word, HostError>;
> 
>     fn pop_stack_dword(&mut self) -> Result<[Word; 2], HostError>;
> 
>     fn pop_stack_vec(&mut self, len: usize) -> Result<Vec<Felt>, HostError>;
> }
> ```

On this point are you proposing that the `Host` remains "dumb" and that the decorator logic remains in the virtual machine?  I think this is quite an important question for the design of the host.  This question leads to - do we want the implementer to be able to modify the decorator logic or should they only be able to modify the advice provider logic.  To help illustrate this with a concrete example - lets take the `miden-base` `SignalHandler` as an example.  Should the `SignalHandler` live within the `AdviceProvider` and the signal be emitted as part of the advice injector methods?  The other option is that the `AdviceProvider` has no knowledge of the `SignalHandler` and if the implementer wants to introduce this component they would do this at the `Host` level and modify the decorator logic to emit a signal when the desired decorator is invoked. 

> For type-safety, we can use the `HostResult` with this as well - though, I'm not sure if it will be easy to work with as for every call to `execute_host_function` we'll need to handle the `match` statement. But maybe it won't be too bad.

I think this question heavily depends on whether the decorator handler logic lives in the host or the vm.  If it lives in the host then the decorators will have direct access to the advice provider. As such the cross section (number of invocations of `HostFunctions`) of the vm <-> Host interface becomes small (5/6 invocations).  In these places we could use something like
```rust
    let variant = if let HostFunctionResult::Word(word) = result {
        word
    } else {
        unreachable!()
    };
``` 

as the client knows the expected return type from the host function.  This isn't very elegant or clean but given the small cross section maybe its fine. 

If the decorator logic remains in the vm then the cross section becomes significantly larger and may result in unwieldily code.

**Comment by bobbinth** [comment]
> On this point are you proposing that the `Host` remains "dumb" and that the decorator logic remains in the virtual machine?

No - the decorator logic still lives in the `Host`. In my example, the decorator logic for `HdwordToMap` advice injector is defined in `insert_hdword_into_adv_map` method implemented on the `AdviceProvider` trait. The only thing the VM does is pass the advice injector to the host via the `execute_host_function()` method.

The above is true for our default implementation of the `Host` interface, but overall, the `Host` interface does not specify how decorators should be handled (e.g., another implementation may put all handling into the host and not bother with the advice provider at all).

I think there are two questions here:
1. How can we re-define handling of certain decorators.
2. How can we intercept calls to certain decorators.

For the first one, if we want to change how `HdwordToMap` we'd provide our own implementation of `AdviceProvider` and override `insert_into_map` function.

For the second one, we could do the same, or we could provide a different implementation of the host. For example:
```Rust
struct MyHost<A, D> {
    adv_provider: A,
    debug_handler: D,
}

impl<A: AdviceProvider, D: DebugHandler> Host for MyHost<A, D> {
   
    fn execute_host_function<S: ProcessState>(
        &mut self,
        process: &S,
        function: HostFunction,
    ) -> Result<usize, HostError> {
        match function {
            HostFunction::Advice(injector) => match injector {
                AdviceInjector::HdwordToMap { domain: _ } => {
                    // do something custom here
                    self.adv_provider.apply_injector(process, injector)
                }
                _ => self.adv_provider.apply_injector(process, injector)
            },
            HostFunction::Debug(info) => self.debug_handler.process_info(process, info),
        }
    }

    fn pop_stack(&mut self) -> Result<Felt, HostError> {
        self.adv_provider.pop_stack()
    }

    fn pop_stack_word(&mut self) -> Result<Word, HostError> {
        self.adv_provider.pop_stack_word()
    }

    fn pop_stack_dword(&mut self) -> Result<[Word; 2], HostError> {
        self.adv_provider.pop_stack_dword()
    }

    fn pop_stack_vec(&mut self, len: usize) -> Result<Vec<Felt>, HostError> {
        self.adv_provider.pop_stack_vec()
    }
}
```

The latter approach may be a bit cleaner, especially if we move interception logic into a stand-alone function. Signal handler could be introduced in a similar way. For example:

```Rust
struct MyHost<A, D> {
    adv_provider: A,
    debug_handler: D,
}

impl<A: AdviceProvider, D: DebugHandler> Host for MyHost<A, D> {
   
    fn execute_host_function<S: ProcessState>(
        &mut self,
        process: &S,
        function: HostFunction,
    ) -> Result<usize, HostError> {
        match function {
            HostFunction::Advice(injector) => {
                self.handle_signal(injector.into());
                self.adv_provider.apply_injector(process, injector)
            },
            HostFunction::Debug(info) => self.debug_handler.process_info(process, info),
        }
    }

    fn pop_stack(&mut self) -> Result<Felt, HostError> {
        self.adv_provider.pop_stack()
    }

    fn pop_stack_word(&mut self) -> Result<Word, HostError> {
        self.adv_provider.pop_stack_word()
    }

    fn pop_stack_dword(&mut self) -> Result<[Word; 2], HostError> {
        self.adv_provider.pop_stack_dword()
    }

    fn pop_stack_vec(&mut self, len: usize) -> Result<Vec<Felt>, HostError> {
        self.adv_provider.pop_stack_vec()
    }
}

impl<A: AdviceProvider, D: DebugHandler> MyHost<A, D> {
    fn handle_signal(&mut self, signal: Signal) {
        // do something custom here
    }
}
```

The above code assumes that there is a `From<AdviceInjector>` implemented on `Signal`.

**Comment by frisitano** [comment]
Got it! Yep this makes sense.  One additional question for the `AdviceProvider`.  We currently have the decorator logic modularised (advice map injectors, advice stack injectors and smt have their own rust modules) do we want to maintain that modularity by putting this logic into their own traits and then combining them using a supertrait or are we happy with all of the decorators being part of the `AdviceProvider` trait.  We could modularise it as follows:

```rust
// this would be the current `AdviceProvider` trait renamed
// this is where the required methods would be specified
trait AdviceProviderBase {
    ...
}

trait AdviceMapInjectors: AdviceProviderBase {
   // default implementations here
   ...
}

trait AdviceStackInjectors: AdviceProviderBase {
   // default implementations here
   ...
}

trait SmtInjectors: AdviceProviderBase {
   // default implementations here
   ...
}

// this is the super trait
trait AdviceProvider: AdviceProviderBase + AdviceMapInjectors + AdviceStackInjectors + SmtInjectors {
   ...
}
``` 

We would then have to implement these on the `DefaultHost`:

```rust
pub struct DefaultHost<A> {
    adv_provider: A,
}

impl<A: AdviceProviderBase> AdviceProviderBase for DefaultHost<A> {
   // the implementer would implement required methods here
  ...
}

impl<A: AdviceProviderBase> AdviceStackInjectors for DefaultHost<A> {
    // this could remain empty or overrides would be implemented here
}

impl<A: AdviceProviderBase> AdviceMapInjectors for DefaultHost<A> {
    // this could remain empty or overrides would be implemented here
}

impl<A: AdviceProviderBase> SmtInjectors for DefaultHost<A> {
   // this could remain empty of overrides would be implemented here
}


impl<A: AdviceProviderBase> AdviceProvider for DefaultHost<A> {
  // this would remain empty
}
```

Whilst this is arguably more modular I think the burden it places on the implementer is a little too high and complicates the implementation.  If we have all advice stack injectors as part of the `AdviceProvider` trait then it significantly reduces the burden on implementing new `AdviceProvider`s and as such I think the reduction in modularity is justified in this case to make it clearer and easier to implement.

**Comment by bobbinth** [comment]
I would prefer a single `AdviceProvider` trait as well. One potential way to modularize the logic is to have separate modules (with pure functions), and then method from the trait just call these functions. This would allow the code to live in different modules, and wouldn't impose extra burden on the implementers of the `AdviceProvider`.

**Comment by frisitano** [comment]
This is a proposed design for the migration of the proposed (and current) implementation which uses a `usize` return type for host functions to a solution which uses a `HostResult` enum return type.  To support this implementation I am thinking we introduce the concept of an `AdviceExtractor` enum (similar to an `AdviceInjector`) except it is responsible for extracting information from the `Host`.  The enum would look something like:

```rust
pub enum AdviceExtractor {
    PopStack,
    PopStackWord,
    PopStackDWord,
    GetMerklePath{ root: Word, depth: Felt, index: Felt },
}
```

We would introduce the typed `HostResult` enum as follows:

```rust
pub enum HostResult {
    MerklePath(MerklePath),
    DoubleWord([Word; 2]),
    Word(Word),
    Element(Felt),
    Unit,
}
```

The `Host` interface would then become:

```rust
pub trait Host {
    fn execute_host_function<S: ProcessState>(
        &mut self,
        process: &S,
        function: &HostFunction,
    ) -> Result<HostResult, ExecutionError>;
}
```

**Comment by bobbinth** [comment]
I think this works. It is probably a bit more cumbersome to use than calling methods directly - but not by much.

One thing to double check is that we don't have something that is both an "injector" and an "extractor" at the same time.

**Comment by frisitano** [comment]
> I think this works. It is probably a bit more cumbersome to use than calling methods directly - but not by much.

Agreed it is slightly more cumbersome, however given the small cross section / number of invocations of these functions in the VM maybe its not such a big deal.

I have implemented both implementation, one with the `usize` return and the other with `AdviceExtractor` and `HostResult` type. Lets review and compare and see which is best to proceed with:

`usize` implementation: https://github.com/0xPolygonMiden/miden-vm/pull/1082
`HostResult` implementation: https://github.com/0xPolygonMiden/miden-vm/pull/1085

> One thing to double check is that we don't have something that is both an "injector" and an "extractor" at the same time.

There is one instance where we may consider an operation to be an "injector" and "extractor" at the same time.  This is the [`UpdateMerkleNode`](https://github.com/0xPolygonMiden/miden-vm/blob/frisitano-host-refactor-host-result/core/src/operations/decorators/advice.rs#L77-L95) injector which returns a `HostResult::MerklePath(path)` where the path is the Merkle path of the new node in the tree. However, I don't think this is necessarily a problem.

**Comment by bobbinth** [comment]
Closed by #1082.
